syntax = "proto3";

package marketplace;

// Enum for item categories
enum Category {
  ELECTRONICS = 0;
  FASHION = 1;
  OTHERS = 2;
  ANY = 3;
}

// Messages for items, requests, and responses
message Item {
  int64 id = 1;
  string name = 2;
  Category category = 3;
  int32 quantity = 4;
  string description = 5;
  string seller_address = 6;
  float price = 7;
  float rating = 8;
}

message RegisterSellerRequest {
  string ip_port = 1;
  string uuid = 2;
}

message OperationResponse {
  bool success = 1;
  string message = 2;
}

message ItemOperationRequest {
  string uuid = 1;
  Item item = 2;
}

message UpdateItemRequest {
  string uuid = 1;
  int64 id = 2;
  int32 quantity = 3;
  float price = 4;
}

message DeleteItemRequest {
  string uuid = 1;
  int64 id = 2;
}

message DisplaySellerItemsRequest {
  string uuid = 1;
}

message SearchRequest {
  string name = 1;
  Category category = 2;
}

message SearchResponse {
  repeated Item items = 1;
}

message DisplaySellerItemsResponse {
  repeated Item items = 1;
}
message BuyItemRequest {
  int64 item_id = 1;
  int32 quantity = 2;
  string buyer_address = 3;
}

message WishlistRequest {
  int64 item_id = 1;
  string buyer_address = 2;
}

message RateItemRequest {
  int64 item_id = 1;
  int32 rating = 2; // Assuming rating is an integer from 1 to 5
  string buyer_address = 3;
}
message NotificationRequest {
  string buyer_address = 1;
  string uuid = 2;
}

message NotificationResponse {
  repeated string messages = 1;
}

// The service definition for marketplace operations
service Marketplace {
  rpc RegisterSeller(RegisterSellerRequest) returns (OperationResponse) {}
  rpc AddItem(ItemOperationRequest) returns (OperationResponse) {}
  rpc UpdateItem(UpdateItemRequest) returns (OperationResponse) {}
  rpc DeleteItem(DeleteItemRequest) returns (OperationResponse) {}
  rpc DisplaySellerItems(DisplaySellerItemsRequest) returns (DisplaySellerItemsResponse) {}
  rpc SearchItems(SearchRequest) returns (SearchResponse) {}
  rpc BuyItem(BuyItemRequest) returns (OperationResponse) {}
  rpc AddToWishList(WishlistRequest) returns (OperationResponse) {}
  rpc RateItem(RateItemRequest) returns (OperationResponse) {}
  rpc FetchNotifications(NotificationRequest) returns (NotificationResponse) {}
  rpc FetchSellerNotifications(NotificationRequest) returns (NotificationResponse);
}
